<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 2 AlphaSimR | Introductory genetics through tilapia breeding simulations</title>
  <meta name="description" content="Chapter 2 AlphaSimR | Introductory genetics through tilapia breeding simulations" />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 2 AlphaSimR | Introductory genetics through tilapia breeding simulations" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 AlphaSimR | Introductory genetics through tilapia breeding simulations" />
  
  
  

<meta name="author" content="PURA acuicultura" />


<meta name="date" content="2021-01-04" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>
<link rel="next" href="recombination.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>


<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Introductory genetics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="alphasimr.html"><a href="alphasimr.html"><i class="fa fa-check"></i><b>2</b> AlphaSimR</a><ul>
<li class="chapter" data-level="2.1" data-path="alphasimr.html"><a href="alphasimr.html#introduction-1"><i class="fa fa-check"></i><b>2.1</b> Introduction</a></li>
<li class="chapter" data-level="2.2" data-path="alphasimr.html"><a href="alphasimr.html#creating-founder-haplotypes"><i class="fa fa-check"></i><b>2.2</b> Creating founder haplotypes</a></li>
<li class="chapter" data-level="2.3" data-path="alphasimr.html"><a href="alphasimr.html#setting-simulation-parameters"><i class="fa fa-check"></i><b>2.3</b> Setting simulation parameters</a></li>
<li class="chapter" data-level="2.4" data-path="alphasimr.html"><a href="alphasimr.html#adding-traits-to-a-breeding-simulation"><i class="fa fa-check"></i><b>2.4</b> Adding traits to a breeding simulation</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="recombination.html"><a href="recombination.html"><i class="fa fa-check"></i><b>3</b> Recombination</a><ul>
<li class="chapter" data-level="3.1" data-path="recombination.html"><a href="recombination.html#introduction-2"><i class="fa fa-check"></i><b>3.1</b> Introduction</a></li>
<li class="chapter" data-level="3.2" data-path="recombination.html"><a href="recombination.html#create-parent-generation"><i class="fa fa-check"></i><b>3.2</b> Create parent generation</a></li>
<li class="chapter" data-level="3.3" data-path="recombination.html"><a href="recombination.html#make-a-full-sib-family"><i class="fa fa-check"></i><b>3.3</b> Make a full-sib family</a></li>
<li class="chapter" data-level="3.4" data-path="recombination.html"><a href="recombination.html#haplotypes-of-the-parent-generation"><i class="fa fa-check"></i><b>3.4</b> Haplotypes of the parent generation</a></li>
<li class="chapter" data-level="3.5" data-path="recombination.html"><a href="recombination.html#haplotypes-of-the-offspring-generation"><i class="fa fa-check"></i><b>3.5</b> Haplotypes of the offspring generation</a></li>
<li class="chapter" data-level="3.6" data-path="recombination.html"><a href="recombination.html#assigning-parental-haplotypes-to-offspring-haplotypes"><i class="fa fa-check"></i><b>3.6</b> Assigning parental haplotypes to offspring haplotypes</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="populations.html"><a href="populations.html"><i class="fa fa-check"></i><b>4</b> Populations</a><ul>
<li class="chapter" data-level="4.1" data-path="populations.html"><a href="populations.html#introduction-3"><i class="fa fa-check"></i><b>4.1</b> Introduction</a></li>
<li class="chapter" data-level="4.2" data-path="populations.html"><a href="populations.html#create-a-test-population"><i class="fa fa-check"></i><b>4.2</b> Create a test population</a></li>
<li class="chapter" data-level="4.3" data-path="populations.html"><a href="populations.html#observe-gene-frequencies"><i class="fa fa-check"></i><b>4.3</b> Observe gene frequencies</a></li>
<li class="chapter" data-level="4.4" data-path="populations.html"><a href="populations.html#observing-snp-haplotype-frequencies"><i class="fa fa-check"></i><b>4.4</b> Observing SNP haplotype frequencies</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="physical-mapping-of-genomes.html"><a href="physical-mapping-of-genomes.html"><i class="fa fa-check"></i><b>5</b> Physical mapping of genomes</a><ul>
<li class="chapter" data-level="5.1" data-path="physical-mapping-of-genomes.html"><a href="physical-mapping-of-genomes.html#introduction-4"><i class="fa fa-check"></i><b>5.1</b> Introduction</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="understanding-additive-traits.html"><a href="understanding-additive-traits.html"><i class="fa fa-check"></i><b>6</b> Understanding additive traits</a><ul>
<li class="chapter" data-level="6.1" data-path="understanding-additive-traits.html"><a href="understanding-additive-traits.html#introduction-5"><i class="fa fa-check"></i><b>6.1</b> Introduction</a></li>
<li class="chapter" data-level="6.2" data-path="understanding-additive-traits.html"><a href="understanding-additive-traits.html#simulating-a-population-to-explore-additive-traits"><i class="fa fa-check"></i><b>6.2</b> Simulating a population to explore additive traits</a></li>
<li class="chapter" data-level="6.3" data-path="understanding-additive-traits.html"><a href="understanding-additive-traits.html#breeding-genetic-and-phenotypic-value"><i class="fa fa-check"></i><b>6.3</b> Breeding, genetic, and phenotypic value</a></li>
<li class="chapter" data-level="6.4" data-path="understanding-additive-traits.html"><a href="understanding-additive-traits.html#heritability-of-additive-traits"><i class="fa fa-check"></i><b>6.4</b> Heritability of additive traits</a></li>
<li class="chapter" data-level="6.5" data-path="understanding-additive-traits.html"><a href="understanding-additive-traits.html#individual-qtl-effects"><i class="fa fa-check"></i><b>6.5</b> Individual QTL effects</a></li>
<li class="chapter" data-level="6.6" data-path="understanding-additive-traits.html"><a href="understanding-additive-traits.html#genome-wide-association-of-additive-traits"><i class="fa fa-check"></i><b>6.6</b> Genome wide association of additive traits</a></li>
<li class="chapter" data-level="6.7" data-path="understanding-additive-traits.html"><a href="understanding-additive-traits.html#qtl-mapping"><i class="fa fa-check"></i><b>6.7</b> QTL mapping</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="simulating-a-tilapia-like-genome.html"><a href="simulating-a-tilapia-like-genome.html"><i class="fa fa-check"></i><b>7</b> Simulating a tilapia like genome</a><ul>
<li class="chapter" data-level="7.1" data-path="simulating-a-tilapia-like-genome.html"><a href="simulating-a-tilapia-like-genome.html#introduction-6"><i class="fa fa-check"></i><b>7.1</b> Introduction</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introductory genetics through tilapia breeding simulations</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="alphasimr" class="section level1">
<h1><span class="header-section-number">Chapter 2</span> AlphaSimR</h1>
<div id="introduction-1" class="section level2">
<h2><span class="header-section-number">2.1</span> Introduction</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="alphasimr.html#cb1-1"></a><span class="kw">library</span>(AlphaSimR)</span>
<span id="cb1-2"><a href="alphasimr.html#cb1-2"></a><span class="kw">library</span>(tidyverse)</span></code></pre></div>
</div>
<div id="creating-founder-haplotypes" class="section level2">
<h2><span class="header-section-number">2.2</span> Creating founder haplotypes</h2>
<p>The first step in developing a population for simulation is the definition of founder haplotypes. In creating founder haplotypes, the genomic architecture of the to be simulated organism is defined. In AlphaSimR, haplotypes are binary strings of 0’s and 1’s. Strings are arranged in a pre-defined number of chromosomes and to a pre-defined length in units of segregation sites. Given that the custom design of haplotypes can be a bit overwhelming to first time users, the function <strong>quickHaplo</strong> can initially be used to get started. <strong>quickHaplo</strong> allows the user to easily define the number of individuals, number of chromosomes per individual, number of segregation sites per chromosome, ploidy level, and the genetic length of chromosomes in units of Morgans. An option for creating inbred founders is also possible.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="alphasimr.html#cb2-1"></a>pop_haplos &lt;-<span class="st"> </span><span class="kw">quickHaplo</span>(<span class="dt">nInd =</span> <span class="dv">10</span>,      <span class="co"># number of individuals</span></span>
<span id="cb2-2"><a href="alphasimr.html#cb2-2"></a>                         <span class="dt">nChr =</span> <span class="dv">1</span>,       <span class="co"># number of chromosomes</span></span>
<span id="cb2-3"><a href="alphasimr.html#cb2-3"></a>                         <span class="dt">segSites =</span> <span class="dv">10</span>,  <span class="co"># number of segregation sites</span></span>
<span id="cb2-4"><a href="alphasimr.html#cb2-4"></a>                         <span class="dt">genLen =</span> <span class="dv">1</span>,     <span class="co"># genetic length in Morgans</span></span>
<span id="cb2-5"><a href="alphasimr.html#cb2-5"></a>                         <span class="dt">ploidy =</span> 2L,    <span class="co"># ploidy level of organism</span></span>
<span id="cb2-6"><a href="alphasimr.html#cb2-6"></a>                         <span class="dt">inbred =</span> <span class="ot">FALSE</span>) <span class="co"># are founders inbred?</span></span></code></pre></div>
<p><strong>quickHaplo</strong> saves haplotype information as a <strong>MapPop</strong> class object. Each slot can be accessed using @ symbol, with list slots requiring additional designations in <strong>[]</strong>. The <strong>MapPop</strong> file provides a record of important genome characteristics. Knowing how to access these characteristics is important to secondary analyses of breeding simulations.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="alphasimr.html#cb3-1"></a>pop_haplos<span class="op">@</span>genMap[<span class="dv">1</span>] <span class="co"># a vector of segregation sites locations (M) per [chromosome]</span></span>
<span id="cb3-2"><a href="alphasimr.html#cb3-2"></a>pop_haplos<span class="op">@</span>centromere[<span class="dv">1</span>] <span class="co"># a vector of centromere location (M) per [chromosome]</span></span>
<span id="cb3-3"><a href="alphasimr.html#cb3-3"></a>pop_haplos<span class="op">@</span>nInd <span class="co"># the number of individuals</span></span>
<span id="cb3-4"><a href="alphasimr.html#cb3-4"></a>pop_haplos<span class="op">@</span>nChr <span class="co"># the number of chromosomes</span></span>
<span id="cb3-5"><a href="alphasimr.html#cb3-5"></a>pop_haplos<span class="op">@</span>ploidy <span class="co"># the ploidy of individuals</span></span>
<span id="cb3-6"><a href="alphasimr.html#cb3-6"></a>pop_haplos<span class="op">@</span>nLoci[<span class="dv">1</span>] <span class="co"># a vector of loci number per [chromosome]</span></span>
<span id="cb3-7"><a href="alphasimr.html#cb3-7"></a>pop_haplos<span class="op">@</span>geno[<span class="dv">1</span>] <span class="co"># matrices of haplotypes per [individual]</span></span></code></pre></div>
<p>Definition of haplotypes is covered in greater detail throughout chapter 1. Chromosome number, segregation sites, haplotypes maps, and centromeres are central to physical genome mapping. Genetic length is covered in a thorough investigation of recombination in AlphaSimR.</p>
</div>
<div id="setting-simulation-parameters" class="section level2">
<h2><span class="header-section-number">2.3</span> Setting simulation parameters</h2>
<p>After defining the founder population haplotypes, simulation parameters are set. As their name implies, simulation parameters define the conditions through which a simulation operates. AlphaSimR possesses a myriad of simulation parameters – far more than can be covered here. A review of the most important parameters to the breeding simulations covered throughout the book are illustrated here. The first step in defining simulation parameters is to create a <strong>SimParam</strong> object. After this is done, simulation parameters can be added sequentially.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="alphasimr.html#cb4-1"></a>SP =<span class="st"> </span>SimParam<span class="op">$</span><span class="kw">new</span>(pop_haplos) <span class="co"># create a variable for storing new simulation parameters</span></span></code></pre></div>
</div>
<div id="adding-traits-to-a-breeding-simulation" class="section level2">
<h2><span class="header-section-number">2.4</span> Adding traits to a breeding simulation</h2>
<p>When paired in diploid individuals, the binary haplotypes become genotypes. At a given locus an individual can be 00, 01, or 11. While it is possible to simply simulate and evaluate genotypic changes in AlphaSimR, most breeding simulations aim to also explore phenotypic change. The first step in simulating phenotypic change is define linkages between the genotypes of specific loci and traits of interest for simulation. Traits of any complexity can be added.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="alphasimr.html#cb5-1"></a>SP<span class="op">$</span><span class="kw">addTraitA</span>(<span class="dt">nQtlPerChr =</span> <span class="dv">3</span>, <span class="co"># number of QTL for trait</span></span>
<span id="cb5-2"><a href="alphasimr.html#cb5-2"></a>             <span class="dt">mean =</span> <span class="dv">454</span>, <span class="co"># mean genetic value of the trait</span></span>
<span id="cb5-3"><a href="alphasimr.html#cb5-3"></a>             <span class="dt">var =</span> <span class="dv">50</span>, <span class="co"># variance of trait in population</span></span>
<span id="cb5-4"><a href="alphasimr.html#cb5-4"></a>             <span class="dt">corA =</span> <span class="ot">NULL</span>, <span class="co"># matrix of correlations between additive effects</span></span>
<span id="cb5-5"><a href="alphasimr.html#cb5-5"></a>             <span class="dt">gamma =</span> <span class="ot">FALSE</span>, <span class="co"># to use a gamma distribution in place of a normal distribution</span></span>
<span id="cb5-6"><a href="alphasimr.html#cb5-6"></a>             <span class="dt">shape =</span> <span class="dv">1</span>, <span class="co"># shape parameter for gamma distribution only</span></span>
<span id="cb5-7"><a href="alphasimr.html#cb5-7"></a>             <span class="dt">force =</span> <span class="ot">FALSE</span>) <span class="co"># keep false until this is understood!</span></span></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="recombination.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/pura-acuicultura/Introductory-genetics/edit/master/02-alphaSimR.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/pura-acuicultura/Introductory-genetics/blob/master/02-alphaSimR.Rmd",
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
