<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Recombination | Introductory genetics through tilapia breeding simulations</title>
  <meta name="description" content="Chapter 3 Recombination | Introductory genetics through tilapia breeding simulations" />
  <meta name="generator" content="bookdown 0.21 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Recombination | Introductory genetics through tilapia breeding simulations" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Recombination | Introductory genetics through tilapia breeding simulations" />
  
  
  

<meta name="author" content="PURA acuicultura" />


<meta name="date" content="2021-01-05" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="a-brief-introduction-to-alphasimr.html"/>
<link rel="next" href="populations.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>


<style type="text/css">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Introductory genetics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="a-brief-introduction-to-alphasimr.html"><a href="a-brief-introduction-to-alphasimr.html"><i class="fa fa-check"></i><b>2</b> A brief introduction to AlphaSimR</a><ul>
<li class="chapter" data-level="2.1" data-path="a-brief-introduction-to-alphasimr.html"><a href="a-brief-introduction-to-alphasimr.html#introduction-1"><i class="fa fa-check"></i><b>2.1</b> Introduction</a></li>
<li class="chapter" data-level="2.2" data-path="a-brief-introduction-to-alphasimr.html"><a href="a-brief-introduction-to-alphasimr.html#creating-founder-haplotypes"><i class="fa fa-check"></i><b>2.2</b> Creating founder haplotypes</a></li>
<li class="chapter" data-level="2.3" data-path="a-brief-introduction-to-alphasimr.html"><a href="a-brief-introduction-to-alphasimr.html#setting-simulation-parameters"><i class="fa fa-check"></i><b>2.3</b> Setting simulation parameters</a><ul>
<li class="chapter" data-level="2.3.1" data-path="a-brief-introduction-to-alphasimr.html"><a href="a-brief-introduction-to-alphasimr.html#adding-traits-to-a-breeding-simulation"><i class="fa fa-check"></i><b>2.3.1</b> Adding traits to a breeding simulation</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="recombination.html"><a href="recombination.html"><i class="fa fa-check"></i><b>3</b> Recombination</a><ul>
<li class="chapter" data-level="3.1" data-path="recombination.html"><a href="recombination.html#introduction-2"><i class="fa fa-check"></i><b>3.1</b> Introduction</a></li>
<li class="chapter" data-level="3.2" data-path="recombination.html"><a href="recombination.html#create-parent-generation"><i class="fa fa-check"></i><b>3.2</b> Create parent generation</a></li>
<li class="chapter" data-level="3.3" data-path="recombination.html"><a href="recombination.html#make-a-full-sib-family"><i class="fa fa-check"></i><b>3.3</b> Make a full-sib family</a></li>
<li class="chapter" data-level="3.4" data-path="recombination.html"><a href="recombination.html#haplotypes-of-the-parent-generation"><i class="fa fa-check"></i><b>3.4</b> Haplotypes of the parent generation</a></li>
<li class="chapter" data-level="3.5" data-path="recombination.html"><a href="recombination.html#haplotypes-of-the-offspring-generation"><i class="fa fa-check"></i><b>3.5</b> Haplotypes of the offspring generation</a></li>
<li class="chapter" data-level="3.6" data-path="recombination.html"><a href="recombination.html#assigning-parental-haplotypes-to-offspring-haplotypes"><i class="fa fa-check"></i><b>3.6</b> Assigning parental haplotypes to offspring haplotypes</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="populations.html"><a href="populations.html"><i class="fa fa-check"></i><b>4</b> Populations</a><ul>
<li class="chapter" data-level="4.1" data-path="populations.html"><a href="populations.html#introduction-3"><i class="fa fa-check"></i><b>4.1</b> Introduction</a></li>
<li class="chapter" data-level="4.2" data-path="populations.html"><a href="populations.html#create-a-test-population"><i class="fa fa-check"></i><b>4.2</b> Create a test population</a></li>
<li class="chapter" data-level="4.3" data-path="populations.html"><a href="populations.html#observe-gene-frequencies"><i class="fa fa-check"></i><b>4.3</b> Observe gene frequencies</a></li>
<li class="chapter" data-level="4.4" data-path="populations.html"><a href="populations.html#observing-snp-haplotype-frequencies"><i class="fa fa-check"></i><b>4.4</b> Observing SNP haplotype frequencies</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="physical-mapping-of-genomes.html"><a href="physical-mapping-of-genomes.html"><i class="fa fa-check"></i><b>5</b> Physical mapping of genomes</a><ul>
<li class="chapter" data-level="5.1" data-path="physical-mapping-of-genomes.html"><a href="physical-mapping-of-genomes.html#introduction-4"><i class="fa fa-check"></i><b>5.1</b> Introduction</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="understanding-additive-traits.html"><a href="understanding-additive-traits.html"><i class="fa fa-check"></i><b>6</b> Understanding additive traits</a><ul>
<li class="chapter" data-level="6.1" data-path="understanding-additive-traits.html"><a href="understanding-additive-traits.html#introduction-5"><i class="fa fa-check"></i><b>6.1</b> Introduction</a></li>
<li class="chapter" data-level="6.2" data-path="understanding-additive-traits.html"><a href="understanding-additive-traits.html#simulating-a-population-to-explore-additive-traits"><i class="fa fa-check"></i><b>6.2</b> Simulating a population to explore additive traits</a></li>
<li class="chapter" data-level="6.3" data-path="understanding-additive-traits.html"><a href="understanding-additive-traits.html#breeding-genetic-and-phenotypic-value"><i class="fa fa-check"></i><b>6.3</b> Breeding, genetic, and phenotypic value</a></li>
<li class="chapter" data-level="6.4" data-path="understanding-additive-traits.html"><a href="understanding-additive-traits.html#heritability-of-additive-traits"><i class="fa fa-check"></i><b>6.4</b> Heritability of additive traits</a></li>
<li class="chapter" data-level="6.5" data-path="understanding-additive-traits.html"><a href="understanding-additive-traits.html#individual-qtl-effects"><i class="fa fa-check"></i><b>6.5</b> Individual QTL effects</a></li>
<li class="chapter" data-level="6.6" data-path="understanding-additive-traits.html"><a href="understanding-additive-traits.html#genome-wide-association-of-additive-traits"><i class="fa fa-check"></i><b>6.6</b> Genome wide association of additive traits</a></li>
<li class="chapter" data-level="6.7" data-path="understanding-additive-traits.html"><a href="understanding-additive-traits.html#qtl-mapping"><i class="fa fa-check"></i><b>6.7</b> QTL mapping</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="simulating-a-tilapia-like-genome.html"><a href="simulating-a-tilapia-like-genome.html"><i class="fa fa-check"></i><b>7</b> Simulating a tilapia like genome</a><ul>
<li class="chapter" data-level="7.1" data-path="simulating-a-tilapia-like-genome.html"><a href="simulating-a-tilapia-like-genome.html#introduction-6"><i class="fa fa-check"></i><b>7.1</b> Introduction</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introductory genetics through tilapia breeding simulations</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="recombination" class="section level1">
<h1><span class="header-section-number">Chapter 3</span> Recombination</h1>
<div id="introduction-2" class="section level2">
<h2><span class="header-section-number">3.1</span> Introduction</h2>
<p>In exploring recombination, <strong>AlphaSimR</strong> is used to create parents and make full sib families. In the assessment of recombination a few other libraries are used: <strong>Matrix</strong>, <strong>tidyverse</strong>, and <strong>gridExtra</strong>.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="recombination.html#cb8-1"></a><span class="kw">library</span>(AlphaSimR)</span>
<span id="cb8-2"><a href="recombination.html#cb8-2"></a><span class="kw">library</span>(Matrix)</span>
<span id="cb8-3"><a href="recombination.html#cb8-3"></a><span class="kw">library</span>(tidyverse)</span>
<span id="cb8-4"><a href="recombination.html#cb8-4"></a><span class="kw">library</span>(gridExtra)</span></code></pre></div>
</div>
<div id="create-parent-generation" class="section level2">
<h2><span class="header-section-number">3.2</span> Create parent generation</h2>
<p>To understand how recombination works in <strong>AlphaSimR</strong> we start with a simple model that can be built upon later. We start by using the <strong>quickHaplo</strong> command to create two parent fish, each diploid with a single chromosome having 10 segregation sites. The length of the chromosome in Morgans is a key determinant of recombination. We will start with a value of 1 cM and then explore how chromosome length impacts recombination.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="recombination.html#cb9-1"></a>founderPop &lt;-<span class="st"> </span><span class="kw">quickHaplo</span>(<span class="dt">nInd =</span> <span class="dv">2</span>, <span class="dt">nChr =</span> <span class="dv">1</span>, <span class="dt">segSites =</span> <span class="dv">10</span>, <span class="dt">genLen =</span> <span class="dv">1</span>, <span class="dt">ploidy =</span> 2L, <span class="dt">inbred =</span> <span class="ot">FALSE</span>)</span>
<span id="cb9-2"><a href="recombination.html#cb9-2"></a>SP &lt;-<span class="st"> </span>SimParam<span class="op">$</span><span class="kw">new</span>(founderPop)</span>
<span id="cb9-3"><a href="recombination.html#cb9-3"></a>pop1 &lt;-<span class="st"> </span><span class="kw">newPop</span>(founderPop, <span class="dt">simParam =</span> SP)</span></code></pre></div>
</div>
<div id="make-a-full-sib-family" class="section level2">
<h2><span class="header-section-number">3.3</span> Make a full-sib family</h2>
<p>Meiotic recombination in parent fish can be observed in the haplotypes of their offspring. We use the <strong>makeCross</strong> command to simulate the cross. The command requires a cross plan in the form of a two column matrix. The rows of the matrix correspond to the cross and the columns indicate the parents. We make only a single cross of parents 1 and 2. Consequently, our cross plan matrix is straightforward.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="recombination.html#cb10-1"></a>crossPlan &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="dt">nrow =</span> <span class="dv">1</span>, <span class="dt">ncol =</span> <span class="dv">2</span>)</span>
<span id="cb10-2"><a href="recombination.html#cb10-2"></a>crossPlan</span></code></pre></div>
<pre><code>##      [,1] [,2]
## [1,]    1    2</code></pre>
<p>Aside from the cross plan, we only need to specify the number of progeny to create. High fecundity in fish is statistically convenient with regards to progeny analysis. We start with a relatively modest 10 progeny but can increase the progeny number to 100 or more later.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="recombination.html#cb12-1"></a>pop2 &lt;-<span class="st"> </span><span class="kw">makeCross</span>(<span class="dt">pop =</span> pop1, <span class="dt">crossPlan =</span> crossPlan, <span class="dt">nProgeny =</span> <span class="dv">10</span>, <span class="dt">simParam =</span> SP)</span></code></pre></div>
</div>
<div id="haplotypes-of-the-parent-generation" class="section level2">
<h2><span class="header-section-number">3.4</span> Haplotypes of the parent generation</h2>
<p>In performing a simulation we have access to information that would be difficult or near impossible to achieve in the real world. Phased haplotypes are one such piece of information, that is also important in the analysis of recombination. We discuss how phased haplotypes can be predicted from actual SNP data in a later chapter. Since our goal here is to understand recombination in <strong>AlphaSimR</strong>, we simply pull them from the simulation.</p>
<p>We start with the parent haplotypes. We pull the haplotypes using the <strong>pullSegSiteHaplo</strong> command. A matrix of binary haplotypes is returned by parent (m = 4) and loci (n = 10). More specifically, row 1 corresponds to parent 1 / haplotype 1, row 2 to parent 1 / haplotype 2, row 3 to parent 2 / haplotype 1, and row 4 to parent 2 / haplotype 2. As such, column 1, referred to as ‘SITE_1’, corresponds to the binary allele (0/1) present at loci 1 of the four parents. Subsequent columns correspond to the remaining alleles designated by our <strong>founderPop</strong>. Now you can appreciate why we started with a small example.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="recombination.html#cb13-1"></a>par_H &lt;-<span class="st"> </span><span class="kw">pullSegSiteHaplo</span>(<span class="dt">pop =</span> pop1, <span class="dt">haplo =</span> <span class="st">&quot;all&quot;</span>, <span class="dt">chr =</span> <span class="dv">1</span>, <span class="dt">simParam =</span> SP)</span>
<span id="cb13-2"><a href="recombination.html#cb13-2"></a>par_H</span></code></pre></div>
<pre><code>##     SITE_1 SITE_2 SITE_3 SITE_4 SITE_5 SITE_6 SITE_7 SITE_8 SITE_9 SITE_10
## 1_1      1      0      1      1      0      0      1      0      0       1
## 1_2      1      1      1      0      0      1      0      0      1       0
## 2_1      1      0      0      1      1      0      0      0      0       1
## 2_2      1      0      0      1      0      0      0      1      0       0</code></pre>
<p>Here we clean up the matrix for some downstream analysis. If we define matrix dimensions and names with <strong>Pop-Class</strong> objects we won’t have to rewrite the code every time we increase the number of parents or segregation sites. The number of rows, or number of parental haplotypes, can be expressed as the number of parents (<strong><a href="mailto:pop1@nInd" class="email">pop1@nInd</a></strong>) multiplied by their ploidy (<strong><a href="mailto:pop1@ploidy" class="email">pop1@ploidy</a></strong>) . The number of columns, or number of loci, can be directly called as <strong><a href="mailto:pop1@nLoci" class="email">pop1@nLoci</a></strong>. In naming the loci using <strong>1:<a href="mailto:pop1@nLoci" class="email">pop1@nLoci</a></strong> a dynamic vector accommodating any number of loci is employed.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="recombination.html#cb15-1"></a>par_H &lt;-<span class="st"> </span><span class="kw">matrix</span>(par_H,  <span class="dt">nrow =</span> pop1<span class="op">@</span>nInd <span class="op">*</span><span class="st"> </span>pop1<span class="op">@</span>ploidy, <span class="dt">ncol =</span> pop1<span class="op">@</span>nLoci,<span class="dt">byrow =</span> <span class="ot">FALSE</span>)</span>
<span id="cb15-2"><a href="recombination.html#cb15-2"></a><span class="kw">colnames</span>(par_H) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;L&quot;</span>, <span class="dv">1</span><span class="op">:</span>pop1<span class="op">@</span>nLoci, <span class="dt">sep =</span> <span class="st">&quot;.&quot;</span>)</span>
<span id="cb15-3"><a href="recombination.html#cb15-3"></a>par_H</span></code></pre></div>
<pre><code>##      L.1 L.2 L.3 L.4 L.5 L.6 L.7 L.8 L.9 L.10
## [1,]   1   0   1   1   0   0   1   0   0    1
## [2,]   1   1   1   0   0   1   0   0   1    0
## [3,]   1   0   0   1   1   0   0   0   0    1
## [4,]   1   0   0   1   0   0   0   1   0    0</code></pre>
</div>
<div id="haplotypes-of-the-offspring-generation" class="section level2">
<h2><span class="header-section-number">3.5</span> Haplotypes of the offspring generation</h2>
<p>We now make a corresponding matrix of offspring haplotypes applying the exact same code to <strong>pop2</strong>. The first four offspring haplotypes are shown.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="recombination.html#cb17-1"></a>off_H &lt;-<span class="st"> </span><span class="kw">pullSegSiteHaplo</span>(<span class="dt">pop =</span> pop2, <span class="dt">haplo =</span> <span class="st">&quot;all&quot;</span>, <span class="dt">chr =</span> <span class="dv">1</span>, <span class="dt">simParam =</span> SP)</span>
<span id="cb17-2"><a href="recombination.html#cb17-2"></a>off_H &lt;-<span class="st"> </span><span class="kw">matrix</span>(off_H, <span class="dt">nrow =</span> pop2<span class="op">@</span>nInd <span class="op">*</span><span class="st"> </span>pop2<span class="op">@</span>ploidy, <span class="dt">ncol =</span> pop2<span class="op">@</span>nLoci, <span class="dt">byrow =</span> <span class="ot">FALSE</span>)</span>
<span id="cb17-3"><a href="recombination.html#cb17-3"></a><span class="kw">colnames</span>(off_H) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">&quot;L&quot;</span>, <span class="dv">1</span><span class="op">:</span>pop2<span class="op">@</span>nLoci, <span class="dt">sep =</span> <span class="st">&quot;.&quot;</span>)</span>
<span id="cb17-4"><a href="recombination.html#cb17-4"></a><span class="kw">head</span>(off_H, <span class="dv">4</span>)</span></code></pre></div>
<pre><code>##      L.1 L.2 L.3 L.4 L.5 L.6 L.7 L.8 L.9 L.10
## [1,]   1   1   1   0   0   1   0   0   1    0
## [2,]   1   0   0   1   0   0   0   0   0    1
## [3,]   1   1   1   0   0   1   0   0   1    0
## [4,]   1   0   0   1   0   0   0   0   0    1</code></pre>
<p>Although we continue using the matrices in subsequent operations, now is a good time to set up a few qualifiers so we don’t lose track of which each row of loci corresponds to which individual and haplotype. Given that it will form the basis for our analysis of recombination, we will do this just for the matrix of offspring haplotypes. As for our other scripts, we make our qualifiers dynamic by using <strong>Pop-Class</strong> objects. We start by assigning an ascending haplotype ID corresponding to the numbered rows of the offspring haplotype matrix. Offspring and allele IDs are similarly assigned. The qualifiers are compiled into a data.frame. Although we could easily bind the offspring matrix to the qualifiers, a few calculations will first be performed.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="recombination.html#cb19-1"></a>HAP &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span>(pop2<span class="op">@</span>nInd <span class="op">*</span><span class="st"> </span>pop2<span class="op">@</span>ploidy)</span>
<span id="cb19-2"><a href="recombination.html#cb19-2"></a>ID &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="dv">1</span><span class="op">:</span>pop2<span class="op">@</span>nInd, <span class="dt">each =</span> pop2<span class="op">@</span>ploidy)</span>
<span id="cb19-3"><a href="recombination.html#cb19-3"></a>ALE &lt;-<span class="kw">rep</span>(<span class="dv">1</span><span class="op">:</span>pop2<span class="op">@</span>ploidy, pop2<span class="op">@</span>nInd)</span>
<span id="cb19-4"><a href="recombination.html#cb19-4"></a>OFF_Q &lt;-<span class="st"> </span><span class="kw">data.frame</span>(HAP, ID, ALE)</span>
<span id="cb19-5"><a href="recombination.html#cb19-5"></a><span class="kw">head</span>(OFF_Q, <span class="dv">4</span>)</span></code></pre></div>
<pre><code>##   HAP ID ALE
## 1   1  1   1
## 2   2  1   2
## 3   3  2   1
## 4   4  2   2</code></pre>
</div>
<div id="assigning-parental-haplotypes-to-offspring-haplotypes" class="section level2">
<h2><span class="header-section-number">3.6</span> Assigning parental haplotypes to offspring haplotypes</h2>
<p>In assessing recombination we are going to match each offspring haplotype with the parental haplotype it resembles most. We will then compare the parent / offspring haplotypes pairs to determine at which loci recombination has taken place. To do all of this matching and comparing we will use <strong>for loops</strong>. In R programming, <strong>for loops</strong> are generally avoided and replaced with <strong>apply</strong> functions. Here, a <strong>for loop</strong> is shown because it is easier to ‘talk through’. To see <strong>for loop</strong> alternatives, look at the <strong>recomb_report</strong> function code in the Appendix.</p>
<p>The first <strong>for loop</strong> is going to compare offspring haplotypes with parental haplotypes.</p>
<p>We first need to set up a <strong>data.frame</strong> to accept the data generate by the <strong>for loop</strong>.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="recombination.html#cb21-1"></a>par_SEL &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">matrix</span>(<span class="dv">0</span>, <span class="dt">nrow =</span> pop2<span class="op">@</span>nInd <span class="op">*</span><span class="st"> </span>pop2<span class="op">@</span>ploidy, <span class="dt">ncol =</span> pop1<span class="op">@</span>nInd <span class="op">*</span><span class="st"> </span>pop1<span class="op">@</span>ploidy))</span>
<span id="cb21-2"><a href="recombination.html#cb21-2"></a><span class="kw">colnames</span>(par_SEL) &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span>(pop1<span class="op">@</span>nInd <span class="op">*</span><span class="st"> </span>pop1<span class="op">@</span>ploidy)</span>
<span id="cb21-3"><a href="recombination.html#cb21-3"></a></span>
<span id="cb21-4"><a href="recombination.html#cb21-4"></a></span>
<span id="cb21-5"><a href="recombination.html#cb21-5"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span>(founderPop<span class="op">@</span>nInd <span class="op">*</span><span class="st"> </span>founderPop<span class="op">@</span>ploidy)){</span>
<span id="cb21-6"><a href="recombination.html#cb21-6"></a>  par_SEL[,i] &lt;-<span class="st"> </span><span class="kw">rowSums</span>((off_H <span class="op">-</span><span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rep</span>(par_H[i,], </span>
<span id="cb21-7"><a href="recombination.html#cb21-7"></a>                                    pop2<span class="op">@</span>nInd <span class="op">*</span><span class="st"> </span>pop2<span class="op">@</span>ploidy),</span>
<span id="cb21-8"><a href="recombination.html#cb21-8"></a>                                <span class="dt">nrow =</span> pop2<span class="op">@</span>nInd <span class="op">*</span><span class="st"> </span>pop2<span class="op">@</span>ploidy,</span>
<span id="cb21-9"><a href="recombination.html#cb21-9"></a>                                <span class="dt">ncol =</span> pop2<span class="op">@</span>nLoci,</span>
<span id="cb21-10"><a href="recombination.html#cb21-10"></a>                                <span class="dt">byrow =</span> <span class="ot">TRUE</span>))<span class="op">^</span><span class="dv">2</span>) </span>
<span id="cb21-11"><a href="recombination.html#cb21-11"></a>}</span></code></pre></div>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="a-brief-introduction-to-alphasimr.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="populations.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/pura-acuicultura/Introductory-genetics/edit/master/03-recombination.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/pura-acuicultura/Introductory-genetics/blob/master/03-recombination.Rmd",
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
